server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -288,6 +288,7 @@ public void prepareForTranslogOperations(boolean fileBasedRecovery, int totalTra
         ActionListener.completeWith(listener, () -> {
             state().getTranslog().totalOperations(totalTranslogOps);
             indexShard().openEngineAndSkipTranslogRecovery();
+            indexShard.trimOperationOfPreviousPrimaryTerms(indexShard.getLastSyncedGlobalCheckpoint());
```

Comment:
With this, operation-based peer recovery becomes destructive (all ops above global checkpoint are effectively erased). I think this will hurt us if we want to transition to a non-destructive model.
We should do this trimming only later in the recovery, when we have caught up with the primary up to endingSeqNo but before we potentially mark the shard as in-sync